lib/zlib.js

```diff
@@ -225,230 +108,330 @@ function zlibBufferSync(engine, buffer) {
   return engine._processChunk(buffer, flushFlag);
 }
 
-// generic zlib
-// minimal 2-byte header
-function Deflate(opts) {
-  if (!(this instanceof Deflate)) return new Deflate(opts);
-  Zlib.call(this, opts, constants.DEFLATE);
+// the Zlib class they all inherit from
+// This thing manages the queue of requests, and returns
+// true or false if there is anything in the queue when
+// you call the .write() method.
+
+function zlibOnError(message, errno) {
+  // there is no way to cleanly recover.
+  // continuing only obscures problems.
+  _close(this);
+  this._hadError = true;
+
+  var error = new Error(message);
+  error.errno = errno;
+  error.code = codes[errno];
+  this.emit('error', error);
 }
 
-function Inflate(opts) {
-  if (!(this instanceof Inflate)) return new Inflate(opts);
-  Zlib.call(this, opts, constants.INFLATE);
+function flushCallback(level, strategy, callback) {
+  assert(this._handle, 'zlib binding closed');
+  this._handle.params(level, strategy);
+  if (!this._hadError) {
+    this._level = level;
+    this._strategy = strategy;
+    if (callback) callback();
+  }
 }
 
+class Zlib extends Transform {
+  constructor(opts, mode) {
+    opts = opts || {};
+    super(opts);
 
-// gzip - bigger header, same deflate compression
-function Gzip(opts) {
-  if (!(this instanceof Gzip)) return new Gzip(opts);
-  Zlib.call(this, opts, constants.GZIP);
-}
+    this._opts = opts;
+    this._chunkSize = opts.chunkSize || constants.Z_DEFAULT_CHUNK;
 
-function Gunzip(opts) {
-  if (!(this instanceof Gunzip)) return new Gunzip(opts);
-  Zlib.call(this, opts, constants.GUNZIP);
-}
+    if (opts.flush && isInvalidFlushFlag(opts.flush)) {
+      throw new RangeError('Invalid flush flag: ' + opts.flush);
+    }
+    if (opts.finishFlush && isInvalidFlushFlag(opts.finishFlush)) {
+      throw new RangeError('Invalid flush flag: ' + opts.finishFlush);
```

Comment:
I'm going to be opening a PR that replaces these with the new internal/errors so wanted to leave these alone for now